src/OpenTelemetry/Metrics/CounterMetricSdkBase.cs

```diff
@@ -59,7 +59,7 @@ internal BoundCounterMetric<T> Bind(LabelSet labelset, bool isShortLived)
             lock (this.bindUnbindLock)
             {
                 var recStatus = isShortLived ? RecordStatus.UpdatePending : RecordStatus.Bound;
-                boundInstrument = this.counterBoundInstruments.GetOrAdd(labelset, this.CreateMetric(recStatus));
+                boundInstrument = this.counterBoundInstruments.GetOrAdd(labelset, (_) => this.CreateMetric(recStatus));
```

Comment:
Pretty sure this still allocates a delegate for the lambda. The ones below aren't using a closure so we could just pass a static ref to avoid that. For the one here we could try calling TryGetValue first and then TryAdd if needed. Not sure which perf hit is worse, double-call or allocation.

Reply:
I agree with your comments.  However, the initial concern was more on the logical side as it relates to calling this.CreateMetric() (and new *MetricSdk()) unnecessarily, rather than the allocation for the delegate and/or closure.  This may become a concern as these "constructors" become more complex and requires additional resources to complete.

As to your concerns regarding lambdas/delegates, I can try doing some profiling of delegates vs methods. (i.e. TryGetValue/TryAdd/etc...)  In my initial google search, it seems the debate is leaning toward use of delegates.

Reply:
After doing some research and running some perf tests on my own, I've come to the following conclusions…

1.	Current PR does not 100% solve the issue of over-allocating a TValue when inserting into a ConcurrentDictionary.  The PR will reduce the probability of this “leakage”.

2.	Use of Lambda expressions with GetOrAdd is more concise (IMHO) vs using TryGetValue/TryAdd/etc…  However, there may be a small performance penalty if lambda expression is “declared”/allocated constantly (as in the case of closures, with older .NET).

3.	I don’t have enough context (of this project) to know what priorities and tradeoffs are valued, and thus, I look forward to feedback from owners and members here.

In more details… in the context of using ConcurrentDictionary<int,int> and my test program inserting and reading 40 Million entries…

1.	GetOrAdd(TKey,TValue) seems to be equivalent of CompareAndExchange operator.  Given current code base, we need to pre-allocate a TValue before calling GetOrAdd and can result in over allocation when race condition inserts TKey before us.  This is also true if we use the delegate/Func/Lambda versions of GetOrAdd, as per remarks in the official documentation.

2.	Performance (CPU) is equivalent when comparing Func<> vs TryGetValue/TryAdd/GetOrAdd.  But, only if we pay special attention to make sure Func<> declarations are minimized (i.e. static Func<>, although, this really should be relegated to the Compiler/Optimizer to handle. Unfortunately, we need to explicitly help it). 

3. or Func<> requiring closures, static Func<> declaration is possible in newer .NET when using the new GetOrAdd(TKey,TArg,TValue) form.

Reply:
Not for this PR but related to ConcurrentDictionary, we've seen several cases where folks hit OOM due to the 2GB limit of ConcurrentDictionary on 64bit machines.

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-5.0#remarks

There are other caveats - e.g. https://github.com/dotnet/runtime/pull/41840#issuecomment-687480893.

Reply:
@victlu 
> I don’t have enough context (of this project) to know what priorities and tradeoffs are valued, and thus, I look forward to feedback from owners and members here.

My perspective on this is that anyone using OpenTelemetry (ipso facto) is sensitive to performance so the library should be as low-impact to the hosting process as possible.

Check out this benchmark:

```csharp
    [MemoryDiagnoser]
    public class ConcurrentDictionaryBenchmarks
    {
        private ConcurrentDictionary<LabelSet, object> instruments;

        [IterationSetup]
        public void IterationSetup()
        {
            this.instruments = new ConcurrentDictionary<LabelSet, object>();
        }

        [Benchmark]
        public void GetOrAdd()
        {
            int closure = 0;

            this.instruments.GetOrAdd(LabelSet.BlankLabelSet, (labelSet) => new object[closure]);

            this.instruments.GetOrAdd(LabelSet.BlankLabelSet, (labelSet) => new object[closure]);
        }

        [Benchmark]
        public void TryAdd()
        {
            int closure = 0;

            if (!this.instruments.TryGetValue(LabelSet.BlankLabelSet, out object value))
            {
                this.instruments.TryAdd(LabelSet.BlankLabelSet, new object[closure]);
            }

            if (!this.instruments.TryGetValue(LabelSet.BlankLabelSet, out value))
            {
                this.instruments.TryAdd(LabelSet.BlankLabelSet, new object[closure]);
            }
        }
    }
```

|   Method |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
| GetOrAdd | 2.732 us | 0.2642 us | 0.7537 us | 2.600 us |     - |     - |     - |     224 B |
|   TryAdd | 2.408 us | 0.2731 us | 0.7658 us | 2.100 us |     - |     - |     - |      72 B |

According to that, doing `TryGetValue`+`TryAdd` is faster and uses less memory. So for me, it's a no-brainer :smile:

I do agree with you the lambda syntax is much nicer, the automatic allocation of a delegate in C# is a tragedy!

To @reyang's point, the code is in a `lock`...
```csharp
            lock (this.bindUnbindLock)
            {
                var recStatus = isShortLived ? RecordStatus.UpdatePending : RecordStatus.Bound;
                boundInstrument = this.counterBoundInstruments.GetOrAdd(labelset, this.CreateMetric(recStatus));
            }
```
...so why the heck are we even using a ConcurrentDictionary? 🤣 I haven't spent much time in the metrics code.

/cc @alanwest @eddynaka

Reply:
@CodeBlanch Thanks for the tip on BenchmarkDotNet!  That is an awesome library.

Your benchmark confirmed my findings as well.  So, if you do this...

        [Benchmark]
        public void GetOrAdd()
        {
            int closure = 0;

            Func<string, int, object> func = (labelSet, closure) => new object[closure];

            this.instruments.GetOrAdd("blank", func, closure);

            this.instruments.GetOrAdd("blank", func, closure);
        }

You get this...

|   Method |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
| GetOrAdd | 2.323 us | 0.2878 us | 0.8258 us | 2.000 us |     - |     - |     - |      72 B |
|   TryAdd | 2.577 us | 0.3651 us | 1.0476 us | 2.000 us |     - |     - |     - |      72 B |

Unfortunately, the GetOrAdd(TKey,TArg,TValue) overload exists only on newer .NET.  Without this form, we would still require allocation for the closure; in which case; TryAdd version is faster.

Reply:
Also, I did not necessarily have to pull out func<> separately.  The Compiler is suppose to optimize that portion.  So...

        [Benchmark]
        public void GetOrAdd()
        {
            int closure = 0;

            //Func<string, int, object> func = (labelSet, closure) => new object[closure];

            this.instruments.GetOrAdd(LabelSet.BlankLabelSet, (l, c) => new object[c], closure);

            this.instruments.GetOrAdd(LabelSet.BlankLabelSet, (l, c) => new object[c], closure);
        }

results in...

|   Method |     Mean |     Error |   StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |---------:|----------:|---------:|---------:|------:|------:|------:|----------:|
| GetOrAdd | 2.597 us | 0.3755 us | 1.101 us | 2.100 us |     - |     - |     - |      72 B |
|   TryAdd | 2.591 us | 0.3495 us | 1.003 us | 2.000 us |     - |     - |     - |      72 B |

Reply:
I think maybe we need to stay focus on the original goal.  Which is to not over-allocate the object we're inserting into the ConcurrentDictionary.  Of which, unfortunately, we only partially addressed. I have strategies for this issue if we remain focus on this.

Also, I understand the whole Metric spec is WIP, and thus, maybe my effort here is premature?

Reply:
Given concern for use of Func<> and support-ability in older .NET, I refactor to using TryGetValue/GetOrAdd instead.  This should optimize (although it does not 100% solve) our over-allocation of dictionary objects.